drivers/at86rf2xx/include/at86rf2xx_aes.h

```diff
@@ -0,0 +1,254 @@
+#define AT86RF2XX_AES_CTRL_AES_REQUEST__START           (0x80)
+
+#define AT86RF2XX_AES_CTRL_AES_MODE__ECB                (0x00)
+#define AT86RF2XX_AES_CTRL_AES_MODE__KEY                (0x10)
+#define AT86RF2XX_AES_CTRL_AES_MODE__CBC                (0x20)
+
+#define AT86RF2XX_AES_CTRL_AES_DIR__ENC                 (0x00)
+#define AT86RF2XX_AES_CTRL_AES_DIR__DEC                 (0x08)
+/** @} */
+
+/**
+ * @name    Layout of register AES_CTRL_MIRROR
+ * @{
+ */
+#define AT86RF2XX_AES_CTRL_MIRROR_MASK__AES_REQUEST     (0x80)
+#define AT86RF2XX_AES_CTRL_MIRROR_MASK__AES_MODE        (0x70)
+#define AT86RF2XX_AES_CTRL_MIRROR_MASK__AES_DIR         (0x08)
+
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_REQUEST__IDLE     (0x00)
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_REQUEST__START    (0x80)
+
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_MODE__ECB         (0x00)
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_MODE__KEY         (0x10)
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_MODE__CBC         (0x20)
+
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_DIR__ENC          (0x00)
+#define AT86RF2XX_AES_CTRL_MIRROR_AES_DIR__DEC          (0x08)
+/** @} */
+
+/**
+ * @brief   An AES block
+ *
+ * AES works on blocks of 16 bytes
+ */
+typedef uint8_t aes_block_t[AT86RF2XX_AES_BLOCK_SIZE];
+
+/**
+ * @brief   Read the AES key used for encryption
+ *
+ * @param[in]       dev     Device
+ * @param[out]      key     Buffer to store the key
+ */
+void at86rf2xx_aes_key_read_encrypt(at86rf2xx_t *dev,
+                                    uint8_t key[AT86RF2XX_AES_KEY_LENGTH]);
+
+/**
+ * @brief   Write the AES key used for encryption
+ *
+ * It is important to write the encryption key, before encryption is done
+ *
+ * @param[in]       dev     Device
+ * @param[in]       key     Buffer which stores the key
+ */
+void at86rf2xx_aes_key_write_encrypt(at86rf2xx_t *dev,
+                                     const uint8_t key[AT86RF2XX_AES_KEY_LENGTH]);
+
+/**
+ * @brief   Read the AES key used for decryption
+ *
+ * @param[in]       dev     Device
+ * @param[out]      key     Buffer to store the key
+ */
+void at86rf2xx_aes_key_read_decrypt(at86rf2xx_t *dev,
+                                    uint8_t key[AT86RF2XX_AES_KEY_LENGTH]);
+
+/**
+ * @brief   Write the AES key used for decryption
+ *
+ * It is important to write the decryption key, before decryption is done
+ *
+ * @param[in]       dev     Device
+ * @param[in]       key     Buffer which stores the key
+ */
+void at86rf2xx_aes_key_write_decrypt(at86rf2xx_t *dev,
+                                     const uint8_t key[AT86RF2XX_AES_KEY_LENGTH]);
+
+/**
+ * @brief   Perform AES algorithm and encrypt data blocks
+ *          in @p plain to cipher data blocks, using ECB mode
+ *
+ * @note    The encryption key must have been written before.
+ *
+ * @param[in]       dev     Device
+ * @param[out]      cipher  If not NULL, cipher data blocks
+ * @param[out]      key     If not NULL, decryption key is stored
```

Comment:
What is this used for?

Reply:
You mean why is there the parameter key?
Before you send data to be encrypted, you must send the encryption key. Then after you have requested to perform an AES encryption, the key register will be updated by the transceiver and contain the last round key of the AES algorithm, which is the key to decrypt the obtained cipher data.

Here is the paragraph from the datasheet:

> When  decrypting,  due  to  the  nature  of  AES  algorithm,  the  initial  key  to  be  used  is  not  the same as the one used for encryption, but rather the last round key instead. This last round  key  is  the  content  of  the  key  address  space  stored  after  running  one  full encryption cycle, and must be saved for decryption. If the decryption key has not been saved,  it  has  to  be  recomputed  by  first  running  a  dummy  encryption  (of  an  arbitrary  plaintext)  using  the  original  encryption  key,  then  fetching  the  resulting  round  key  from the key memory, and writing it back into the key memory as the decryption key.

I see I can use `_aes_save_key()` in `at86rf2xx_aes_key_read_encrypt()` and `at86rf2xx_aes_key_read_decrypt()`

Reply:
Then better change the comment to

>  `* @param[out]      key     If not NULL, last round decryption key is stored`

to avoid other people getting as confused as me.

Reply:
yes